backend/src/services/user_session_manager/user_session_manager.py

```diff
@@ -0,0 +1,309 @@
+            job_component_name=session_def.job_component_name,
+            job_scheduler_port=session_def.port,
+            instance_str=effective_instance_str,
+            actual_service_port=actual_service_port,
+            approx_timeout_s=40,
+        )
+        if existing_session_info:
+            session_url = f"http://{existing_session_info.radix_job_name}:{actual_service_port}"
+            LOGGER.info(
+                f"Got existing user session in: {timer.elapsed_ms()}ms ({user_component=}, {instance_str=}, {session_url=})"
+            )
+            return session_url
+
+        LOGGER.debug(
+            f"Unable to get existing user session, starting new session for: {user_component=}, {instance_str=}"
+        )
+        new_session_info = await _create_new_session(
+            session_dir=session_dir,
+            job_component_name=session_def.job_component_name,
+            job_scheduler_port=session_def.port,
+            resource_req=session_def.resource_req,
+            instance_str=effective_instance_str,
+            actual_service_port=actual_service_port,
+            approx_timeout_s=30,
+        )
+
+        if not new_session_info:
+            LOGGER.error(f"Failed to create new user session for: {user_component=}, {instance_str=}")
+            return None
+
+        session_url = f"http://{new_session_info.radix_job_name}:{actual_service_port}"
+        LOGGER.info(
+            f"Created new user session in: {timer.elapsed_ms()}ms ({user_component=}, {instance_str=}, {session_url=})"
+        )
+
+        return session_url
+
+
+# Look for existing session info, possibly waiting for a partially created session to come online
+#
+# Reasons why this function may return None
+#   1. We found the session, but it's not running yet (it's in the process of being created/spinning
+#      up or has got stuck while being created)
+#   2. We did not find the session in the directory, and it needs to be created
+#   3. We found the session and the info says it is running, but our verification probe against
+#      radix or a probe against the service's health endpoint says it's not
+async def _get_info_for_running_session(
+    session_dir: UserSessionDirectory,
+    job_component_name: str,
+    job_scheduler_port: int,
+    instance_str: str,
+    actual_service_port: int,
+    approx_timeout_s: float,
+) -> SessionInfo | None:
+
+    time_counter = TimeCounter(approx_timeout_s)
+    sleep_time_s = 1
+    num_calls = 1
+
+    session_info = session_dir.get_session_info(job_component_name, instance_str)
+    if not session_info:
+        return None
+
+    # Given that we found info on the session and it's not in the running state, we will try and wait for it to come online.
+    # The job/session might be in the process of being created and spinning up, so we will try and wait for it.
+    # We will need to add a timeout to this!!!!!
+    # How much time should we spend here before giving up?
```

Comment:
Is startup time of the docker image neglible here, if so it is a question we could ask Radix? Maybe they have a timeout already before a session/job is marked as failed(-to-provision)?

Reply:
Removed the stale comment and added som more detailed comments on timeouts while we seek the truth either through practical usage or from the Radix team.